### DISCUSSION TITLE
Is it possible to create seperate MaterialGrids for diagonal and off-diagonal elements of the dielectric tensor

### DISCUSSION THREAD

**Discussion by peterropac** [NanoComp/meep#2345] [Q&A]
Hello, 

I am trying to use the adjoint solver in MEEP to optimize the density of two materials which have non-zero off-diagonal elements in their dielectric tensors. The mixing between the off-diagonal elements should happen at different rates than for the diagonal elements. While these rates are related, they are far from being the same. 

I tried solving this problem, by defining a function which takes a position argument converts it to an index in the `MaterialGrid`, gets the weight, does the correct interpolation and returns a mp.Medium object. Using this function the optimization only ran for one iteration, because the value of the objective function didn't change. Is it even possible that this approach would work? Based on other discussions I would say no.

The second idea is to define an array of parameters, which are feed into a function which transformes them into weights of two seperate `MaterialGrid` objects (similar as in the example in  #2235). One `MaterialGrid`  is for the diagonal and the other for the off-diagonal elements of the dielectric tensor. Would it be somehow possible to combine their outputs?

Best!

**Comment by smartalecH** [comment]
> The mixing between the off-diagonal elements should happen at different rates than for the diagonal elements

I'm not sure I understand. Are you saying they have different interpolation functions when mapping from the design parameter ($u$) to the different tensor components? If so, then that's not currently possible, as that particular step is handled under the hood (and just linearly interpolates each tensor component)

Maybe you can talk a bit more about what you're really trying to do and there might be an alternative way to tackle this (without rewriting a bunch of the internal code).

**Reply by peterropac** [reply]
Yes, you got it right that is the bottom line. 

I am designing a liquid crystal optical element, and liquid crystals are birefringent. In order to describe their optical properties, we need to use the dielectric tensor. I wrote code that works for a single component of the dielectric tensor, but that isn't very realistic. I turns out that for very specific conditions and parametrization of the dielectric tensor it is possible to express the diagonal components as `na_ii*v + (1-v)*nb_ii`, where `v = (cos(pi*u))^2` and `u` is the design parameter. Values of `na_ii` and `nb_ii` aren't the same for all diagonal components.  For the off-diagonal elements, a very similar parametrization can be used but this time `v = sin(2*pi*u))/2`. I can convert `u` into `v` by using a `mapping` function as I would if I was doing filtering, projecting, gradient limiting etc., but since `v` for diagonal elements is different from the `v` for off-diagonal elements a single `MaterialGrid `just won't do. 

Ideally, I would be able to define a `material_function` as I described before. For more complex parametrizations of the dielectric tensor the trick in the previous paragraph just won't work.
